### ISSUE TITLE
[Feature request] User-defined functions

### ISSUE THREAD

**Issue by aaaaaa123456789** [gbdev/rgbds#201]
I'm sure that this has been mentioned before, but I've been coming up with a full feature proposal and I'd like to know how feasible it is. What follows is a proposed way of implementing functions, at least in a way I'd consider usable.

The idea is to allow the user to define functions, that would be evaluated in-line when called, similar to the various built-ins. Functions can have one or more arguments, which define variables that are local to the function; they are substituted in the expression. An example of the syntax I have in mind would be:

```
;definition
addnums(first, second) = ({first}) + ({second})

;usage
  ld hl, wTwoPlusTwo
  ld [hl], addnums(2, 2)
```

Functions can be defined with the same name as long as they have a different number of arguments; the proper version of the function would be used when called. For simplicity, all arguments are required.

```
average(one) = {one}
average(one, two) = (({one}) + ({two})) / 2
average(one, two, three) = (({one}) + ({two}) + ({three})) / 3
```

Declaring functions with the same name and number of arguments is not allowed; such declarations would either replace previous ones or cause an error (either works).

Calling built-in functions should obviously be allowed from user-defined functions:

```
linearaddress(label) = (BANK({label}) - 1) * $4000 + ({label})
```

In order to simplify functions like the average one above, varargs functions could be defined. Such functions would be used when there is no suitable non-varargs function (i.e., with the correct number of arguments) to call. For instance:

```
countargs() = 0
countargs(arg, ...) = 1 + countargs({...})

sum() = 0
sum(arg, ...) = ({arg}) + sum({...})

average(...) = sum({...}) / countargs({...})
roundedavg(...) = (2 * sum({...}) + 1) / (2 * countargs({...}))
```

The special symbol `{...}` is replaced by the list of variable arguments. Defining multiple varargs functions with the same name is not allowed; however, they can coexist with non-varargs functions.

As a silly example just to show how they would coexist:

```
fn(a) = {a}
fn(a, b, c) = {c}
fn(a, b, ...) = {b}

  db fn(4) ;4
  db fn(5, 9) ;9
  db fn(2, 6, 3) ;3
  db fn(1, 8, 0, 7) ;8
  ;db fn() is an error because there's no version that takes 0 arguments
```

I hope that this is doable, as it would certainly reduce some repetition and extremely macro-heavy code in some cases. As a simple realistic example, take this macro from Prism (might be slightly different because I'm typing it from memory):

```
;definition
coord: MACRO
  if _NARG > 3
    ld \1, (\4) + (\2) + ((\3) * SCREEN_WIDTH)
  else
    ld \1, wTilemap + (\2) + ((\3) * SCREEN_WIDTH)
  endc
ENDM

hlcoord EQUS "coord hl, "

;usage
  hlcoord 2, 5 ;sets hl to the corresponding tilemap position
```

Which using functions would be a lot cleaner:

```
;definition
coord(x, y, base) = ({base}) + ({x}) + ({y}) * SCREEN_WIDTH
coord(x, y) = coord({x}, {y}, wTilemap)

;usage
  ld hl, coord(2, 5)
```

I await your responses; thanks for reading.

**Comment by Ben10do** [comment]
Interesting! So, if I'm right in saying, the main things that distinguish this from existing macros are:
- The ability to use them as part of an expression (like `EQU`/`EQUS`/`SET`s all can be)
- Named parameters
- Easier overloading (based on number of parameters)

-----------------

I do like that you're going for a functional-style feel, but it bothers me that it's a very different syntax for something that's intuitively similar to macros.

Since I think the latter two features would be also good for macros to have, we could potentially unify their syntaxes, perhaps in a manner similar to this:

```
MyFunction: FUNCTION(foo, bar)   ; Reflects usage as MyFunction(foo, bar)
    \foo + \bar
    ENDF
```
```
MyMacro: MACRO foo, bar          ; Reflects usage as MyMacro foo, bar
    ld a, \foo + \bar            ; Similar to \1, \2
    ENDM
```

The syntax to use them would probably still have to differ, due to the existing difference between macros and built-in functions.

This approach makes functions look procedural, so I don't know if we should do so (and add a `RETURN` statement, and possibly make that implicit for the last line), or only allow a single expression (which would make this rather verbose for a one-liner).

-----------------

Alternately, you could think of it as something more similar to `EQU` or `SET`. Then the `MyFunction() = ...` syntax could be a nicer way of doing `MyFunction() FUNC ...` (similar to `SET`).

But, in order to add a little consistency, I'd consider using `\foo` to access parameters, since it's closer to how you already access parameters in macros:

```
MyFunction(foo, bar) = \foo + \bar
```

-----------------

You can probably tell that I'm just thinking around the issue, and haven't really made any decisions on what I like best. However we implement it though, I do think this would be a useful feature.

Thoughts?

**Comment by AntonioND** [comment]
I prefer the {} version because they are less ambiguous: "\foo1" vs "{foo}1" (see https://github.com/rednex/rgbds/issues/63 ).

In principle it doesn't sound bad. In practice, this is a lot of work. The code that handles macros is quite annoying because it simply cannot be "just parsed", you have to inject the expanded code of the macro...

**Comment by aaaaaa123456789** [comment]
I'd say the functional style is simpler, although I can imagine a few cases in which procedural would be more powerful. That being said, procedural-style functions would require you to actually execute them, so that might increase the burden on your side.

I'm also assuming lazy evaluation here (see the linearaddress example above, which wouldn't work without it because BANK($4444) is invalid), which might be harder to achieve in a procedural way.

That being said, I agree that named parameters and overloads (and even varargs) could be useful for macros, so a unified syntax sounds like a good idea. The function/endf syntax looks a bit long for functional-style one-liners, but the "= replaces func" syntax looks good. I'm not really biased towards any particular syntax, though.

#64 and the numerical lexer abuse in #63 are downright silly, and the assembler should probably just fail and exit if it detects weird behavior like that with functions. I can't see something like that appearing in real code.

By the way, I'm well aware that this is a big feature, so I don't expect it to be finished soon.

**Comment by yenatch** [comment]
Can named args be used on call as well? And can it be multiline?
```asm
    MyFunction (
        foo = 2,
        bar = 1
    )
```

**Comment by ISSOtm** [comment]
Fixed as of 0.4.0, both with the updating of the `mandoc.css` style sheet, and the adding of a custom `rgbds.css` one.

**Comment by ISSOtm** [comment]
One of the things for *sure*: such functions will be "pure". No side effects, only produce an expression.
I would like avoiding the approach macros currently take (re-lexing+parsing a buffer), since that's pretty slow. This brought up a problem we have at all: how do captures work?
```
CONSTANT = 0
DEF foo() => CONSTANT
CONSTANT = 1
db foo()
```
Should this:
- Output 0
- Output 1
- Be an error
?

**Comment by aaaaaa123456789** [comment]
I don't think it should be an error. As for 0 or 1, both approaches are valid, but I'd say 0 is more useful, but 1 is more natural to people used to macros — I'd prefer 0.
~(or make it configurable)~

**Comment by ISSOtm** [comment]
If so, I'll go with 0 for a first implementation, and maybe extend with a "capture" syntax later so 1 can be produced.

**Comment by NieDzejkob** [comment]
If I'm understanding correctly, the behavior "output 1" could create dependency cycles...

**Comment by ISSOtm** [comment]
No, because the right-hand is always evaluated before the left-hand.
The behavior that's been decided on is "output 0", and a syntax akin to C++'s captures may be added in the future to allow the possibility for "output 1".

**Comment by pinobatch** [comment]
For comparison, [`.define` macros in ca65](https://cc65.github.io/doc/ca65.html#ss12.7) behave much like `#define` macros in C preprocessor, except they don't use parentheses when called.

**Comment by ISSOtm** [comment]
Making these work in a useful way would require #619, if only because recursive functions would require the same kind of expression system rewrite that short-circuiting operators would. Therefore, and because I think 0.4.2 has been delayed enough, this feature will be postponed (again) to 0.4.3.

**Comment by ISSOtm** [comment]
Lazy evaluation of symbols is important. Eager evaluation can be forced using naked braces (since #634), but lazy evaluation cannot otherwise be done.

**Comment by Rangi42** [comment]
[Sjasm](http://www.xl2s.tk/sjasmman11.html#s1) has something similar: "text macros with arguments". (Although these are expanded as statements, not expressions, so more like macros with named parameters.)

**Comment by Rangi42** [comment]
Some implementations for common math functions that aren't built into rgbasm:

```
def abs(x) = x < 0 ? -x : x
def abs(x) = x * (1 - 2 * (x < 0))
def abs(x) = x * sgn(x)

def sgn(x) = x < 0 ? -1 : x > 0 ? 1 : 0
def sgn(x) = (x > 0) - (x < 0)
def sgn(x) = x ? x / abs(x) : 0

def sqrt(x) = pow(x, 0.5)
```

**Comment by Rangi42** [comment]
Functions could allow specifying their arguments by name, in case the given order doesn't match the declaration order.

```
def coord(x, y) = x * SCREEN_WIDTH + y

    ld hl, wTilemap + coord(y=5, x=10)
    ld hl, wTilemap + coord(10, 5)
```

**Comment by ISSOtm** [comment]
This would be especially helpful to clarify intent, as it's neither obvious nor widely agreed upon whether X or Y goes first (typical usage is `X, Y`, but e.g. OAM uses `Y, X`)

**Comment by Rangi42** [comment]
Assuming user-defined functions will support recursion, and that we implement a short-circuiting ternary operator, these can be defined as follows:

```
DEF popcnt(x) := x ? (1 + popcnt(x & (x - 1))) : 0
DEF ctz(x) := x ? (x & 1) ? 0 : (1 + ctz(x >> 1)) : 0
DEF clz(x) := x ? ((x & $8000_0000) ? 0 : (1 + clz(x << 1))) : 0
```

**Comment by Rangi42** [comment]
User-defined functions, as currently planned, are single expressions. However, this means that they can't `assert` anything about their arguments or results. It would be helpful to have some new kind of pre/post-condition assert that can work inside function definitions.

My use case is that pokecrystal repeats this pattern for getting an unlabeled SRAM location from its labeled WRAM counterpart:
```asm
ld hl, sPlayerData + (wEventFlags - wPlayerData)
ld hl, sPlayerData + (wHallOfFameCount - wPlayerData)
ld hl, sPlayerData + (wPlayerID - wPlayerData)
; etc
```
What I'd rather do:
```asm
def Saved(label) = sPlayerData + (label - wPlayerData)
ld hl, Saved(wEventFlags)
ld hl, Saved(wHallOfFameCount)
ld hl, Saved(wPlayerID)
; etc
```
However, that would be unsafe to use without a precondition of `label >= wPlayerData && label < wPlayerDataEnd`.

**Comment by aaaaaa123456789** [comment]
A perhaps simple (and probably usable elsewhere) solution is to add an assert function: a function that simply returns 0 (so it's easy to add to an expression by doing `assert(...) + expr`) while creating an assertion inside an expression. This would require the linker to be able to handle that, though.

**Comment by Rangi42** [comment]
Hm, this is the second recent context in which an `assert` function would be useful. (The other was [in pokered](https://github.com/pret/pokered/pull/557#discussion_r2656596307), where we twice want to do `assert COND, "message"` but then use `COND` in some further logic, and having to define a temp variable would be more trouble than it's worth vs just repeating the `COND` expression.)

**Comment by ISSOtm** [comment]
Another way inline assertions would be useful, though more marginal, is `xor a :: assert SOMETHING == 0`, though perhaps that's orthogonal since it's not in an expression.

I don't like `assert()` being a full-blown expression, because the operator between that and the “real” expression is... weird. Perhaps we can enforce syntax like `assert(THING != 0, "awawawaaaaaaa") && STUFF / THING`? (I like this operator in particular because of the implied short-circuiting—yes, I know, 1.0 lacks it, but 2.0 has it.)

**Comment by Rangi42** [comment]
If `assert(numeric condition, string message)` is usable as an inline numeric/Boolean value anywhere at all, it should be a general-purpose one. That's more predictable consistent behavior.

Yes it allows people to do silly things like `xor assert(42, "lol")`, or `assert assert(0, "this prints") == 0, "this doesn't"` but either they're being intentionally silly for fun, or they have non-silly needs that we didn't foresee. Either way, we shouldn't be building in an artificial "`assert(...)` can only be used numerically when followed by a Boolean binary operator" structure.

Although if we're both liking the idea of inline `assert` in *some* capacity, I'll open a separate issue for it.

**Comment by aaaaaa123456789** [comment]
I should probably say that `assert(...) && (expr)` is a bad idea because it intuitively can only evaluate to 0 or 1 (because those are the only possible results of the `&&` operator). And making a special rule just for the `assert` function so that it can give you any result seems like adding a lot of unintuitive behavior. It's _begging_ for a "`\1 && \2` returns 4 instead of 0-1" bug report.

**Comment by Rangi42** [comment]
Let's continue any discussion of inline/result-checkable assertions in #1885.

**Comment by Rangi42** [comment]
WLA-DX has user-defined functions with [`.FUNCTION`](https://wla-dx.readthedocs.io/en/latest/asmdiv.html#function-sum-ab-vara-varb), e.g. `.FUNCTION prod(x, y) (x * y)` and then `prod(2 + 3, 10 / 5)`. (It also has various built-in functions which RGBASM hasn't added because in principle they can be user-defined, eventually, when that's possible. Like `ABS`, `CLAMP`, `MAX`, `MIN`, `RANDOM`, etc.)
